### PR TITLE
User suggestions for bclconverter

### PR DESCRIPTION
**Pull request by sylvinite**
* Can now add undetermined reads to solo samples in a lane. It is still possible to do stupid stuff, but it has a basic sanity check
* Number of failed samples is listed at the end of QC
* Fixed a NaN issue for hiseq
* General tidying up of code

### CODE REVIEW

File: scripts/manage_demux_stats.py [hunk 1]
```diff
@@ -89,76 +90,122 @@ def manipulate_process(demux_process, proc_stats):
...
                 if lane_no == entry["Lane"]:
+                    lane_no = int(lane_no)
                     sample = entry["Sample"]
+                    
+                    if "Lanes to include undetermined" in demux_process.udf:
+                        try:
+                            undet_lanes= re.split('[ ,.]', demux_process.udf["Lanes to include undetermined"])
+                            undet_lanes = [int(i) for i in undet_lanes]
+                        except:
+                            problem_handler("exit", "Unable to typecast included undetermined lanes. Possibly non-number in list")
```
Comment (Galithil) [inline scripts/manage_demux_stats.py]: Why does this block need to be repeated for each entry ? Can't you set the lanes to include undetermined once for all entries ?

File: scripts/manage_demux_stats.py [hunk 2]
```diff
@@ -89,76 +90,122 @@ def manipulate_process(demux_process, proc_stats):
...
+                                try:
+                                    clusterType = None
+                                    if "PF Clusters" in undet:
+                                        clusterType = "PF Clusters"
+                                    else:
+                                        clusterType = "Clusters"
+                                    #Paired runs are divided by two within flowcell parser
+                                    if proc_stats["Paired"]:
+                                        undet_reads = int(undet[clusterType].replace(",",""))*2
+                                        undet_pairs = int(undet[clusterType].replace(",",""))
```
Comment (Galithil) [inline scripts/manage_demux_stats.py]: or just `undet_reads/2`

File: scripts/manage_demux_stats.py [hunk 3]
```diff
@@ -89,76 +90,122 @@ def manipulate_process(demux_process, proc_stats):
...
                         try:
                             target_file.udf["%PF"] = float(entry["% PFClusters"])
                             logger.info("{}% PF".format(str(target_file.udf["%PF"])))
-                            target_file.udf["% One Mismatch Reads (Index)"] = float(entry["% One mismatchbarcode"])
-                            logger.info("{}% One Mismatch Reads (Index)".format(str(target_file.udf["% One Mismatch Reads (Index)"])))
+                            
+                            #["% One mismatchbarcode"] can hold NaN. Treating it as 0.0
+                            if entry["% One mismatchbarcode"] == "NaN":
+                                target_file.udf["% One Mismatch Reads (Index)"] = 0.0
+                                logger.info("'NaN' One Mismatch Reads (Index), treating as {}".format(str(target_file.udf["% One Mismatch Reads (Index)"])))
```
Comment (Galithil) [inline scripts/manage_demux_stats.py]: the plus side of using format is that you don't need to cast stuff to string to include them in a string.

File: scripts/manage_demux_stats.py [hunk 4]
```diff
@@ -179,15 +226,15 @@ def set_sample_values(demux_process, parser_struct, proc_stats):
                                 clusterType = "Clusters"
                             #Paired runs are divided by two within flowcell parser
                             if proc_stats["Paired"]:
-                                target_file.udf["# Reads"] = int(entry[clusterType].replace(",",""))*2
-                                target_file.udf["# Read Pairs"] = int(entry[clusterType].replace(",",""))
+                                target_file.udf["# Reads"] = int(entry[clusterType].replace(",",""))*2 + undet_reads
```
Comment (Galithil) [inline scripts/manage_demux_stats.py]: is under_reads defined in all cases ?

File: scripts/manage_demux_stats.py [hunk 5]
```diff
@@ -222,17 +271,22 @@ def set_sample_values(demux_process, parser_struct, proc_stats):
...
-            #This looks kind of bad in lims, but \n can't be used.
-            problem_handler("warning", "Undemultiplexed reads for lane {} was {} ({})% thus exceeding \
-             defined limit.\t".format(lane_no, undet_lane_reads, found_undet))
-        else:
-            logger.info("Found {} ({}%) undemultiplexed reads for lane {}.".format(undet_lane_reads, found_undet, lane_no))
+        if not undet_included:
+            #If undetermined reads are greater than threshold*reads_in_lane
+            found_undet = round(float(undet_lane_reads)/(assign_lane_reads+undet_lane_reads)*100, 2)
+            if  found_undet > demux_process.udf["Maximum % Undetermined Reads per Lane"]:
+                #This looks kind of bad in lims, but \n can't be used.
```
Comment (Galithil) [inline scripts/manage_demux_stats.py]: If you want to write long and comprehensive errrors, you can also wrtite to a log file and upload it.

**Comment by sylvinite** [comment]
Fixed everything you mentioned. under_reads defaults to 0 when it is not used.

**Pull request merged**